File: templates/typescript/transport/libp2p/Libp2PGossip.ts
/**
 * Libp2PGossip — IGossipProtocol implementation using libp2p GossipSub.
 *
 * Publishes and subscribes to the /borgkit/gossip/1.0.0 topic.
 * Messages are LP-framed JSON GossipMessage objects.
 */
import type { Libp2p }        from 'libp2p';
import type { GossipMessage } from '../../interfaces/IAgentMesh';

export const GOSSIP_TOPIC = '/borgkit/gossip/1.0.0';

type GossipHandler = (msg: GossipMessage) => Promise<void>;

export class Libp2PGossip {
  private readonly node:     Libp2p;
  private readonly handlers: GossipHandler[] = [];

  constructor(node: Libp2p) {
    this.node = node;
  }

  /** Subscribe to the gossip topic and forward messages to registered handlers. */
  subscribe(): void {
    const pubsub = (this.node.services as Record<string, unknown>)['pubsub'] as any;
    if (!pubsub) throw new Error('GossipSub service not enabled on this libp2p node');

    pubsub.subscribe(GOSSIP_TOPIC);
    pubsub.addEventListener('message', (evt: CustomEvent) => {
      if (evt.detail.topic !== GOSSIP_TOPIC) return;
      try {
        const msg = JSON.parse(new TextDecoder().decode(evt.detail.data)) as GossipMessage;
        // Only forward if TTL allows
        if ((msg.ttl ?? 0) > 0) {
          for (const h of this.handlers) h(msg).catch(() => {});
        }
      } catch { /* malformed message — ignore */ }
    });
  }

  /** Publish a GossipMessage to all subscribed peers. */
  async publish(msg: GossipMessage): Promise<void> {
    const pubsub = (this.node.services as Record<string, unknown>)['pubsub'] as any;
    if (!pubsub) return;
    const data = new TextEncoder().encode(JSON.stringify(msg));
    await pubsub.publish(GOSSIP_TOPIC, data);
  }

  /** Register a handler for incoming gossip messages. */
  onMessage(handler: GossipHandler): void {
    this.handlers.push(handler);
  }

  peers(): string[] {
    const pubsub = (this.node.services as Record<string, unknown>)['pubsub'] as any;
    return pubsub?.getSubscribers(GOSSIP_TOPIC)?.map((p: any) => p.toString()) ?? [];
  }
}

File: templates/python/transport/libp2p/libp2p_gossip.py
"""
Libp2PGossipProtocol — IGossipProtocol backed by the Rust FFI GossipSub.
"""
from __future__ import annotations

import json
from typing import Awaitable, Callable, List

from interfaces.iagent_mesh import GossipMessage, IGossipProtocol
from .ffi import BorgkitLibp2P

GossipHandler = Callable[[GossipMessage], Awaitable[None]]


class Libp2PGossipProtocol(IGossipProtocol):
    """
    Publishes gossip messages via the Rust borgkit-libp2p GossipSub.
    Incoming messages arrive via the request handler registered on the FFI node
    (capability == "__gossip_pubsub") and are forwarded to registered handlers.
    """

    def __init__(self, ffi: BorgkitLibp2P) -> None:
        self._ffi      = ffi
        self._handlers: List[GossipHandler] = []
        self._peers:    List[str] = []

    async def broadcast(self, message: GossipMessage) -> None:
        self._ffi.gossip_publish(json.dumps(message.to_dict()))

    async def receive(self, message: GossipMessage) -> None:
        if not message.should_forward:
            return
        for h in self._handlers:
            try:
                await h(message)
            except Exception:
                pass

    def subscribe(self, handler: GossipHandler) -> None:
        self._handlers.append(handler)

    def peers(self) -> List[str]:
        return list(self._peers)

    async def add_peer(self, agent_id: str, endpoint: str) -> None:
        try:
            self._ffi.dial(endpoint)
            self._peers.append(agent_id)
        except ConnectionError:
            pass

    async def remove_peer(self, agent_id: str) -> None:
        self._peers = [p for p in self._peers if p != agent_id]
